components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71498PhenoTips2155DataMigration.java

```diff
@@ -132,264 +202,638 @@ public Object doInHibernate(Session hSession) throws HibernateException, XWikiEx
+                }
+                if (patientReferredBy.containsKey(relativeID)
+                    && patientReferredBy.get(relativeID).size() > 1) {
+                    // this patient is referred to by more than one patient - can't put into more
+                    // than one peigree so don't put into any
+                    continue;
                 }
+                if (FAMSTUDIES_CHILD.equals(relationshipType)) {
+                    numberOfParents++;
+                    if (numberOfParents > 2) {
+                        // can't be a child of more than 2 parents
+                        continue;
+                    }
+                }
+                String owner = this.getBasicPatientData(this.getPatientXDocument(relativeID)).getString(OWNER);
+                if (!StringUtils.equals(owner, probandData.getString(OWNER))) {
+                    // patints are owned by different users
+                    continue;
+                }
+
+                okRelativesList.add(relativeID);
             }
 
-            return null;
+            return okRelativesList;
         }
 
         /**
-         * Returns either existing patient family or relative family document, or a new one. Patient and relatives are
-         * assigned to a new family as members. Returns null if more than one family exists for patient and relatives.
+         * In an existing pedigree finds a node linked to the given patient and adds all given relatives to
+         * pedigree node comments.
          *
-         * @throws XWikiException
-         * @throws QueryException
+         * @param patientID the id of the patient to look for
+         * @param familyID the id of the family with pedigree
+         * @param relativesToAdd the list of relatives
          */
-        private XWikiDocument processPatientWithRelatives(XWikiDocument patientXDocument, String patientFamilyRef,
-            String relativesFamilyRef, Map<String, XWikiDocument> relativesDocList, XWiki xwiki) throws QueryException,
-            XWikiException
+        private void addRelativesAsComment(String patientId, String familyId,
+                Map<String, String> relativesToAdd)
         {
-            XWikiDocument familyXDocument = null;
-            if (relativesFamilyRef == null || patientFamilyRef != null && StringUtils.isNotBlank(relativesFamilyRef)) {
-                this.migrator.logger.debug("More than one family exists for patient and relatives. Patient Id: {}.",
-                    patientXDocument.getId());
+            try {
+                XWikiDocument familyXDocument =
+                        this.context.getWiki().getDocument(this.migrator.resolver.resolve(familyId), this.context);
+
+                BaseObject pedigreeObj = familyXDocument.getXObject(Pedigree.CLASS_REFERENCE);
+                if (pedigreeObj == null) {
+                    this.migrator.logger.error("Error updating existing family: no pedigree object for {}", familyId);
```

Comment:
well, in a way in that other case it can be considered not a big error, since if there is no pedigree there is no thing to migrate, end of story (for that other issue)

In our case, the fact that there supposedly is a pedigree changes how we migrate other data, and it might not get migrated properly if pedigree is broken. So it is OK to have `debug` there and we do need an `error` here